openmls/src/group/core_group/validation.rs

```diff
@@ -264,4 +266,81 @@ impl CoreGroup {
 
         Ok(())
     }
+
+    /// Validate constraints on an external commit. This function implements the following checks:
+    ///  - ValSem240: External Commit, inline Proposals: There MUST be at least one ExternalInit proposal.
+    ///  - ValSem241: External Commit, inline Proposals: There MUST be at most one ExternalInit proposal.
+    ///  - ValSem242: External Commit, inline Proposals: There MUST NOT be any Add proposals.
+    ///  - ValSem243: External Commit, inline Proposals: There MUST NOT be any Update proposals.
+    ///  - ValSem244: External Commit, inline Remove Proposal: The identity and the endpoint_id of the removed
+    ///               leaf are identical to the ones in the path KeyPackage.
+    ///  - ValSem245: External Commit, referenced Proposals: There MUST NOT be any ExternalInit proposals.
+    pub(crate) fn validate_external_commit(
+        &self,
+        proposal_queue: &ProposalQueue,
+        path_key_package_option: Option<&KeyPackage>,
+    ) -> Result<(), CoreGroupError> {
+        let mut external_init_proposals =
+            proposal_queue.filtered_by_type(ProposalType::ExternalInit);
+        // ValSem240: External Commit, inline Proposals: There MUST be at least one ExternalInit proposal.
+        if let Some(external_init_proposal) = external_init_proposals.next() {
+            // ValSem245: External Commit, referenced Proposals: There MUST NOT be any ExternalInit proposals.
+            if external_init_proposal.proposal_or_ref_type() == ProposalOrRefType::Reference {
+                return Err(ExternalCommitValidationError::ReferencedExternalInitProposal.into());
+            }
+        } else {
+            return Err(ExternalCommitValidationError::NoExternalInitProposals.into());
+        };
+
+        // ValSem241: External Commit, inline Proposals: There MUST be at most one ExternalInit proposal.
+        if let Some(additional_external_init_proposal) = external_init_proposals.next() {
+            // ValSem245: External Commit, referenced Proposals: There MUST NOT be any ExternalInit proposals.
+            if additional_external_init_proposal.proposal_or_ref_type()
+                == ProposalOrRefType::Reference
+            {
+                return Err(ExternalCommitValidationError::ReferencedExternalInitProposal.into());
+            } else {
+                return Err(ExternalCommitValidationError::MultipleExternalInitProposals.into());
+            }
+        }
+
+        let add_proposals = proposal_queue.filtered_by_type(ProposalType::Add);
+        for proposal in add_proposals {
+            // ValSem242: External Commit, inline Proposals: There MUST NOT be any Add proposals.
+            if proposal.proposal_or_ref_type() == ProposalOrRefType::Proposal {
```

Comment:
The spec isn't specific about inline proposals here. I guess it's more of a spec problem, but I'm not sure we should allow Add proposals by reference. The same is true for the Update and Remove proposals below.

Reply:
I'm not sure I understand. I think the spec is pretty specific. There is point point about proposals by value (i.e. inline), which is the following:
> * The proposals included by value in an External Commit MUST meet the following conditions:
>   * There MUST be a single ExternalInit proposal
>   * There MUST NOT be any Add or Update proposalsIf a Remove proposal is present, then the credential and endpoint_id of the removed leaf MUST be the same as the corresponding values in the Add KeyPackage.

Similarly, it's explicit about proposals by reference, where it states the following:
> * The proposals included by reference in an External Commit MUST meet the following conditions:
>   * There MUST NOT be any ExternalInit proposals

So I think the checks I'm proposing here line up exactly with what the spec says. Please correct me if I'm wrong. If we want to do more or less, I prefer we change the spec first and then act according to the spec changes.

Reply:
Thanks for the explanation, I missed the last part. I guess it makes sense the way you implemented it.